### PR TITLE
Refactor e2e node selection

### PR DESCRIPTION
**Pull request by msau42**
**What type of PR is this?**
/kind cleanup
/kind flake

**What this PR does / why we need it**:
Pass `e2epod.NodeSelection` directly into storage testsuites so that it gets consistently applied. Also fix `volume.TestConfig` to also use `e2epod.NodeSelection`.

**Which issue(s) this PR fixes**:

Fixes #
Addresses #87855

**Special notes for your reviewer**:
First commit is from https://github.com/kubernetes/kubernetes/pull/88054

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by pohly** [comment]
/retest

From https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/88059/pull-kubernetes-conformance-kind-ipv6-parallel/1227460936009256960/ 
```
[Fail] [k8s.io] Kubelet when scheduling a busybox Pod with hostAliases [It] should write entries to /etc/hosts [LinuxOnly] [NodeConformance] [Conformance] 
```

**Review by pohly — CHANGES_REQUESTED** [review]
@msau42 Can you also include the patch below in your PR? It's another location where currently `pod.Spec.NodeName` is set directly. I ran into this today because late binding does not work when `pod.Spec.NodeName` is set, presumably because the volume binder is never called by the scheduler for pods which have already been scheduled manually. Is that a known limitation of late binding? I've not seen it described before.

```
diff --git a/test/e2e/storage/csi_mock_volume.go b/test/e2e/storage/csi_mock_volume.go
index 2ff9a9c5f77..3937f4fe73a 100644
--- a/test/e2e/storage/csi_mock_volume.go
+++ b/test/e2e/storage/csi_mock_volume.go
@@ -720,12 +720,7 @@ func startPausePodWithVolumeSource(cs clientset.Interface, volumeSource v1.Volum
                },
        }
 
-       if node.Name != "" {
-               pod.Spec.NodeName = node.Name
-       }
-       if len(node.Selector) != 0 {
-               pod.Spec.NodeSelector = node.Selector
-       }
+       e2epod.SetNodeSelection(pod, node)
 
        return cs.CoreV1().Pods(ns).Create(context.TODO(), pod, metav1.CreateOptions{})
 }
```

**Comment by pohly** [comment]
> Can you also include the patch below in your PR?

Or shall we merge this PR now as all tests are now clean and I'll submit that patch separately?

**Comment by pohly** [comment]
OTOH, there seem to be some other places that might need to be fixed:
```
$ git grep 'Spec.NodeName = ' test
test/e2e/apps/statefulset.go:                   ss.Spec.Template.Spec.NodeName = node.Name
test/e2e/framework/pods.go:     pod.Spec.NodeName = TestContext.NodeName
test/e2e/network/firewall.go:                   pod.Spec.NodeName = nodeName
test/e2e/network/no_snat.go:    pod.Spec.NodeName = nodename
test/e2e/network/no_snat.go:    pod.Spec.NodeName = nodename
test/e2e/network/service.go:                    pod.Spec.NodeName = nodeName
test/e2e/network/service.go:                                    rc.Spec.Template.Spec.NodeName = endpointNodeName
test/e2e/node/security_context.go:      pod.Spec.NodeName = foundPod.Spec.NodeName
test/e2e/storage/nfs_persistent_volume-disruptive.go:   pod.Spec.NodeName = nodeName
test/e2e_node/device_plugin_test.go:                    dp.Spec.NodeName = framework.TestContext.NodeName
```

**Comment by msau42** [comment]
Yes, there are a ton of other places that need to be fixed. I was mainly focused on the storage framework tests first because those generate more test cases than the others and are more likely to flake.

The CSI mock tests I was planning to work on separately too because I had more questions on the overall test flow.

File: test/e2e/framework/volume/fixtures.go [hunk 1]
```diff
@@ -128,12 +128,8 @@ type TestConfig struct {
 	// Wait for the pod to terminate successfully
 	// False indicates that the pod is long running
 	WaitForCompletion bool
-	// ServerNodeName is the spec.nodeName to run server pod on.  Default is any node.
-	ServerNodeName string
```
Comment (davidz627) [inline test/e2e/framework/volume/fixtures.go]: Do we need to still have a seperate `ServerNodeSelection`? suspicious that this disappeared

File: test/e2e/storage/persistent_volumes-local.go [hunk 2]
```diff
@@ -988,7 +988,8 @@ func makeLocalPodWithNodeName(config *localTestConfig, volume *localTestVolume,
 	if pod == nil {
 		return
 	}
-	pod.Spec.NodeName = nodeName
+
+	e2epod.SetNodeAffinity(pod, nodeName)
```
Comment (davidz627) [inline test/e2e/storage/persistent_volumes-local.go]: not gonna use the fancy `NodeSelection` stuff here?

File: test/e2e/framework/volume/fixtures.go [hunk 3]
```diff
@@ -297,8 +293,6 @@ func startVolumeServer(client clientset.Interface, config TestConfig) *v1.Pod {
 			},
 			Volumes:       volumes,
 			RestartPolicy: restartPolicy,
-			NodeName:      config.ServerNodeName,
```
Comment (davidz627) [inline test/e2e/framework/volume/fixtures.go]: the only place where `ServerNodeName` was used seems to be here and it's just deleted and not replaced by anything 🤔

**Review by davidz627 — COMMENTED** [review]
mostly LGTM

File: test/e2e/framework/volume/fixtures.go (hunk 3 above)
Comment (msau42) [reply test/e2e/framework/volume/fixtures.go]: correct, nothing is using it, so I just removed it.

File: test/e2e/storage/persistent_volumes-local.go (hunk 2 above)
Comment (msau42) [reply test/e2e/storage/persistent_volumes-local.go]: not at the moment. The local storage tests are specific to single nodes.

**Pull request merged**